src/fileutils.hpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2019, The Bifrost Authors. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * * Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * * Neither the name of The Bifrost Authors nor the names of its
+ *   contributors may be used to endorse or promote products derived
+ *   from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <sys/file.h>  // For flock
+#include <sys/stat.h>  // For fstat
+#include <sys/types.h> // For getpid
+#include <unistd.h>    // For getpid
+#include <pwd.h>       // For getpwuid
+#include <system_error>
+
+#if defined __APPLE__ && __APPLE__
+#include <sys/sysctl.h>
+#endif
+
+inline std::string get_home_dir(void) {
+	const char *homedir;
+  if ((homedir = getenv("HOME")) == NULL) {
+    homedir = getpwuid(getuid())->pw_dir;
+  }
+	return std::string(homedir);
+}
+inline void make_dir(std::string path, int perms=775) {
+	if( std::system(("mkdir -p "+path+" -m "+std::to_string(perms)).c_str()) ) {
+		throw std::runtime_error("Failed to create path: "+path);
+	}
+}
+inline void remove_all(std::string path) {
+	if( std::system(("rm -rf "+path).c_str()) ) {
+		throw std::runtime_error("Failed to remove all: "+path);
+	}
+}
+inline void remove_dir(std::string path) {
+	if( std::system(("rmdir "+path+" 2> /dev/null").c_str()) ) {
+		throw std::runtime_error("Failed to remove dir: "+path);
+	}
+}
+inline void remove_file(std::string path) {
+	if( std::system(("rm -f "+path).c_str()) ) {
+		throw std::runtime_error("Failed to remove file: "+path);
+	}
+}
+inline bool file_exists(std::string path) {
+    struct stat s;
+    return !(stat(path.c_str(), &s) == -1
+	         && errno == ENOENT);
+}
+bool process_exists(pid_t pid) {
```

Comment:
Although this function is long, it either needs to be marked `inline` like the others or moved out into a single `.cpp`. I'm getting a link error because it's duplicated in `map.o` and `proclog.o`:

```
map.o: In function `process_exists(int)':
/home/qblocks/bifrost/src/fileutils.hpp:74: multiple definition of `process_exists(int)'
proclog.o:/home/qblocks/bifrost/src/fileutils.hpp:74: first defined here
```

Will try a fix locally and see what happens next…

Reply:
I see now that most of the functionality now in `fileutils.hpp` was originally in `proclog.o` and not inline. If you like, I can try to figure out how to keep it in a (possibly new) `.cpp` but refer to it from both places. (Rather than inline.)

Reply:
Yes, it was in `proclog.cpp` before but I wanted to use it in `map.cpp` as well.  `process_exists` also used to be a lot simpler when it was linux-only.  The mac support is what caused it to balloon.